File: 0260-single-number-iii/0260-single-number-iii.py
class Solution:
    def singleNumber(self, nums):
        xori = 0
        a = 0
        b = 0
        for num in nums:
            xori ^= num
        val = xori & -xori  # Isolate the rightmost set bit
        for num in nums:
            if val & num:
                a ^= num
            else:
                b ^= num
        return [a, b]

File: 0260-single-number-iii/0260-single-number-iii.cpp
class Solution {
public:
    vector<int> singleNumber(vector<int>& nums) {
        int xori = 0;
        int a = 0, b = 0;
        for (auto it : nums) xori ^= it;
        unsigned int val = xori & -(unsigned int)xori;  // Isolate the rightmost set bit using unsigned negation
        for (auto it : nums) {
            if (val & it) a ^= it;
            else b ^= it;
        }
        return {a, b};
    }
};
